system/src/system_power_manager.cpp

```diff
@@ -552,52 +555,205 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+    if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
+    }
+    return;
+  } else if (g_batteryState == BATTERY_STATE_DISCONNECTED) {
+    if (vCell > BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      if (vcellDebounceCount_ == 0) {
+        // It's probably that the battery is attached, request to shorten the monitor period
+        battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+        DBG_PWR("Battery might be attached");
+      }
+      vcellDebounceCount_++;
+      if (vcellDebounceCount_ >= BATTERY_VCELL_DEBOUNCE_COUNT) {
+        goto connected;
+      }
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    }
+    return;
+  }
+
+connected:
+  PMIC power(true);
+  const uint8_t status = power.getSystemStatus();
+  const uint8_t pwrGood = (status >> 2) & 0b01;
+  if ((status & 0x08) || !pwrGood) {
+    DBG_PWR("In DPM mode or power is not good");
+    // It's in DPM mode, the battery is discharging
+    confirmBatteryState(g_batteryState, BATTERY_STATE_DISCHARGING);
+  } else if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+    DBG_PWR("vCell < BATTERY_CONNECTED_VCELL_THRESHOLD");
+    // If battery falls below 2.0v, it's probably detached
+    confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+  } else {
+    // Since charging is disabled, it must be in the not charging shate
+    confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
   }
+}
+
+void PowerManager::deduceBatteryStateChargeEnabled() {
+  DBG_PWR("deduceBatteryStateChargeEnabled()");
+  PMIC power(true);
 
+  if (g_batteryState == BATTERY_STATE_DISCONNECTED && !power.isChargingEnabled() /* MUST not enabled */) {
+    if (millis() - disableChargingTimeStamp_ < BATTERY_DISABLE_CHARGING_SUPRESS_PERIOD) {
+      return;
+    }
+    // At this point, we are still disabling charging, the VCELL can reliably reflect the voltage on VBAT
+    FuelGauge fuel(true);
+    float vCell = fuel.getVCell();
+    DBG_PWR("VCell: %dmV", (uint16_t)(vCell * 1000));
+    if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      DBG_PWR("It's still disconnected.");
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+      return;
+    }
+    DBG_PWR("The battery might be attached");
+    // Fall through
+  }
+
+  handleCharging();
+
+  const uint8_t status = power.getSystemStatus();
+  const uint8_t chargeState = (status >> 4) & 0b11;
+  if (chargeState == 0b11 && g_batteryState != BATTERY_STATE_CHARGED) {
+    batteryStateTransitioningTo(BATTERY_STATE_CHARGED);
+  } else if ((chargeState == 0b01 || chargeState == 0b10)) {
+    batteryStateTransitioningTo(BATTERY_STATE_CHARGING);
+  } else if (chargeState == 0b00) {
+    const uint8_t pwrGood = (status >> 2) & 0b01;
+    if ((status & 0x08) || !pwrGood) {
+      DBG_PWR("In DPM mode or power is not good");
```

Comment:
I think this is a bit confusing when battery charging is enabled and we have a power source connected. We will be reporting a sticky `discharging` state even though we are still charging the battery while at the same time occasionally taking some charge off of it to compensate for spikes. I would probably stick with just checking `pwrGood` here, especially considering that we will not be getting an event from the PMIC whenever DPM state changes.

cc @eberseth

Reply:
> especially considering that we will not be getting an event from the PMIC whenever DPM state changes

That makes sense 👍

Reply:
I am ok with this change but understand that it would prevent DPM changes